File: ui/src/protocol.ts
import type { MeetingStatus, PCMOwner, RuntimeMode } from "./contracts/meetingContract";

export type DuplexMode = "speaker_focus" | "headphone_duplex";

export interface AudioLevelsSnapshot {
  readonly microphone: number;
  readonly physical_output: number;
  readonly mixed: number;
  readonly updated_at_ns: number;
}

export type ServiceProbeStatus = "ok" | "unreachable" | "timeout" | "error";

export interface ServiceInfo {
  readonly name: string;
  readonly status: ServiceProbeStatus;
  readonly url: string;
  readonly target_model?: string | null;
  readonly model_present?: boolean | null;
  readonly workload?: string | null;
  readonly ws_state?: string | null;
  readonly reconnect_count?: number | null;
  readonly last_event_age_ms?: number | null;
  readonly dropped_chunks?: number | null;
  readonly gap_count?: number | null;
}

export interface ServicesResponse {
  readonly services: ServiceInfo[];
  readonly diagnostics?: unknown;
  readonly network_scope?: "local" | "network";
}

export interface RuntimeStateSnapshot {
  readonly mode: RuntimeMode;
  readonly pcm_owner: PCMOwner;
  readonly active_meeting_id?: string | null;
  readonly meeting_state?: MeetingStatus | null;
  readonly meeting_started_at?: string | null;
  readonly pipeline: string;
  readonly subtitle: string;
  readonly storage?: string;
  readonly mic_muted: boolean;
  readonly runtime_revision: number;
  readonly audio_levels?: AudioLevelsSnapshot;
  readonly persona?: string | null;
  readonly voice?: string;
  readonly duplex_mode?: DuplexMode;
  readonly session_started_at?: string | null;
  readonly degraded_reason?: string | null;
  readonly capabilities?: {
    readonly inner_os_enabled: boolean;
    readonly inner_os_analysis_enabled: boolean;
    readonly inner_os_channel: "loopback_only";
    readonly diarization_overlay_enabled?: boolean;
  };
}

export type ControlCommand =
  | { readonly cmd: "clear_context" }
  | { readonly cmd: "clear_subtitles" }
  | { readonly cmd: "stop_session" }
  | { readonly cmd: "restart" }
  | { readonly cmd: "set_persona"; readonly prompt: string }
  | { readonly cmd: "set_voice"; readonly voice: string }
  | { readonly cmd: "set_duplex_mode"; readonly mode: DuplexMode }
  | { readonly cmd: "set_mic_muted"; readonly muted: boolean }
  | { readonly cmd: "start_meeting"; readonly title?: string; readonly max_speakers?: number; readonly contract_version?: "1" }
  | { readonly cmd: "end_meeting"; readonly meeting_id?: string; readonly contract_version?: "1" }
  | { readonly cmd: "start_assistant"; readonly contract_version?: "1" }
  | { readonly cmd: "start_subtitles"; readonly contract_version?: "1" }
  | { readonly cmd: "stop_active_mode"; readonly contract_version?: "1" }
  | { readonly cmd: "send_text"; readonly text: string };

export interface CommandResponse {
  readonly contract_version?: "1";
  readonly request_id: string;
  readonly cmd: string;
  readonly ok: boolean;
  readonly state: RuntimeStateSnapshot;
  readonly error_code?: string | null;
  readonly message?: string | null;
  readonly error?: {
    readonly code: string;
    readonly message: string;
    readonly request_id?: string;
    readonly details?: Record<string, unknown>;
  } | null;
}

export function isServicesResponse(value: unknown): value is ServicesResponse {
  return isRecord(value)
    && Array.isArray(value.services)
    && value.services.every(isServiceInfo)
    && (value.network_scope === undefined
      || value.network_scope === "local"
      || value.network_scope === "network");
}

function isServiceInfo(value: unknown): value is ServiceInfo {
  return isRecord(value)
    && typeof value.name === "string"
    && isServiceProbeStatus(value.status)
    && typeof value.url === "string"
    && isOptionalString(value.target_model)
    && isOptionalBoolean(value.model_present)
    && isOptionalString(value.workload)
    && isOptionalString(value.ws_state)
    && isOptionalNonNegativeInteger(value.reconnect_count)
    && isOptionalNonNegativeInteger(value.last_event_age_ms)
    && isOptionalNonNegativeInteger(value.dropped_chunks)
    && isOptionalNonNegativeInteger(value.gap_count);
}

function isServiceProbeStatus(value: unknown): value is ServiceProbeStatus {
  return value === "ok"
    || value === "unreachable"
    || value === "timeout"
    || value === "error";
}

function isOptionalString(value: unknown): value is string | null | undefined {
  return value === undefined || value === null || typeof value === "string";
}

function isOptionalBoolean(value: unknown): value is boolean | null | undefined {
  return value === undefined || value === null || typeof value === "boolean";
}

function isOptionalNonNegativeInteger(value: unknown): value is number | null | undefined {
  return value === undefined
    || value === null
    || (typeof value === "number" && Number.isInteger(value) && value >= 0);
}

export function isRuntimeState(value: unknown): value is RuntimeStateSnapshot {
  if (!isRecord(value)) return false;
  return isRuntimeMode(value.mode)
    && isPCMOwner(value.pcm_owner)
    && typeof value.runtime_revision === "number"
    && Number.isInteger(value.runtime_revision)
    && value.runtime_revision >= 0
    && typeof value.pipeline === "string"
    && typeof value.subtitle === "string"
    && typeof value.mic_muted === "boolean"
    && (value.audio_levels === undefined || isAudioLevels(value.audio_levels))
    && (value.active_meeting_id === undefined || typeof value.active_meeting_id === "string" || value.active_meeting_id === null)
    && (value.meeting_state === undefined || isMeetingStatus(value.meeting_state) || value.meeting_state === null)
    && (value.meeting_started_at === undefined || typeof value.meeting_started_at === "string" || value.meeting_started_at === null)
    && (value.storage === undefined || typeof value.storage === "string")
    && (value.persona === undefined || typeof value.persona === "string" || value.persona === null)
    && (value.voice === undefined || typeof value.voice === "string")
    && (value.duplex_mode === undefined || value.duplex_mode === "speaker_focus" || value.duplex_mode === "headphone_duplex")
    && (value.session_started_at === undefined || typeof value.session_started_at === "string" || value.session_started_at === null)
    && (value.degraded_reason === undefined || typeof value.degraded_reason === "string" || value.degraded_reason === null)
    && (value.capabilities === undefined || isRuntimeCapabilities(value.capabilities));
}

function isAudioLevels(value: unknown): value is AudioLevelsSnapshot {
  return isRecord(value)
    && isNormalizedLevel(value.microphone)
    && isNormalizedLevel(value.physical_output)
    && isNormalizedLevel(value.mixed)
    && typeof value.updated_at_ns === "number"
    && Number.isInteger(value.updated_at_ns)
    && value.updated_at_ns >= 0;
}

function isNormalizedLevel(value: unknown): value is number {
  return typeof value === "number"
    && Number.isFinite(value)
    && value >= 0
    && value <= 1;
}

function isRuntimeCapabilities(value: unknown): value is RuntimeStateSnapshot["capabilities"] {
  return isRecord(value)
    && typeof value.inner_os_enabled === "boolean"
    && typeof value.inner_os_analysis_enabled === "boolean"
    && value.inner_os_channel === "loopback_only"
    && (value.diarization_overlay_enabled === undefined
      || typeof value.diarization_overlay_enabled === "boolean");
}

function isRuntimeMode(value: unknown): value is RuntimeMode {
  return value === "assistant" || value === "subtitles" || value === "meeting" || value === "idle";
}

function isPCMOwner(value: unknown): value is PCMOwner {
  return value === "assistant" || value === "subtitles" || value === "meeting" || value === "none";
}

function isMeetingStatus(value: unknown): value is MeetingStatus {
  return value === "recording"
    || value === "finalizing"
    || value === "completed"
    || value === "interrupted"
    || value === "storage_error";
}

export function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

File: src/voice_realtime/ui/protocol.py
"""Voice Studio 控制 WebSocket 的严格消息契约。"""

from __future__ import annotations

from collections.abc import Mapping
from enum import StrEnum
from typing import Annotated, Any, Literal

from pydantic import BaseModel, ConfigDict, Field, StringConstraints, TypeAdapter

from voice_realtime.interaction.types import DuplexMode as DuplexMode
from voice_realtime.meeting.models import MeetingStatus, PCMOwner, RuntimeMode, StorageHealth

RequestId = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, max_length=64)]
ShortText = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1, max_length=128)]
PersonaText = Annotated[
    str,
    StringConstraints(strip_whitespace=True, min_length=1, max_length=4000),
]
MeetingTitle = Annotated[
    str,
    StringConstraints(strip_whitespace=True, min_length=1, max_length=200),
]
MeetingId = Annotated[
    str,
    StringConstraints(strip_whitespace=True, min_length=1, max_length=64),
]


class CommandBase(BaseModel):
    """所有控制命令的公共字段。"""

    model_config = ConfigDict(extra="forbid")

    request_id: RequestId


class ClearContextCommand(CommandBase):
    cmd: Literal["clear_context"]


class ClearSubtitlesCommand(CommandBase):
    cmd: Literal["clear_subtitles"]


class StopSessionCommand(CommandBase):
    cmd: Literal["stop_session"]


class RestartCommand(CommandBase):
    cmd: Literal["restart"]


class SetPersonaCommand(CommandBase):
    cmd: Literal["set_persona"]
    prompt: PersonaText


class SetVoiceCommand(CommandBase):
    cmd: Literal["set_voice"]
    voice: ShortText


class SetDuplexModeCommand(CommandBase):
    cmd: Literal["set_duplex_mode"]
    mode: DuplexMode


class SetBargeInModeCommand(CommandBase):
    """旧前端命令名；语义与 set_duplex_mode 相同。"""

    cmd: Literal["set_barge_in_mode"]
    mode: DuplexMode


class SetMicMutedCommand(CommandBase):
    cmd: Literal["set_mic_muted"]
    muted: bool


class StartMeetingCommand(CommandBase):
    cmd: Literal["start_meeting"]
    title: MeetingTitle | None = None
    max_speakers: int | None = Field(default=None, ge=1, le=4)
    contract_version: Literal["1"] | None = None


class EndMeetingCommand(CommandBase):
    cmd: Literal["end_meeting"]
    meeting_id: MeetingId | None = None
    contract_version: Literal["1"] | None = None


class StartAssistantCommand(CommandBase):
    cmd: Literal["start_assistant"]
    contract_version: Literal["1"] | None = None


class StartSubtitlesCommand(CommandBase):
    cmd: Literal["start_subtitles"]
    contract_version: Literal["1"] | None = None


class StopActiveModeCommand(CommandBase):
    cmd: Literal["stop_active_mode"]
    contract_version: Literal["1"] | None = None


class SendTextCommand(CommandBase):
    cmd: Literal["send_text"]
    text: PersonaText


ControlCommand = Annotated[
    ClearContextCommand
    | ClearSubtitlesCommand
    | StopSessionCommand
    | RestartCommand
    | SetPersonaCommand
    | SetVoiceCommand
    | SetDuplexModeCommand
    | SetBargeInModeCommand
    | SetMicMutedCommand
    | StartMeetingCommand
    | EndMeetingCommand
    | StartAssistantCommand
    | StartSubtitlesCommand
    | StopActiveModeCommand
    | SendTextCommand,
    Field(discriminator="cmd"),
]
_COMMAND_ADAPTER: TypeAdapter[ControlCommand] = TypeAdapter(ControlCommand)


class RuntimeStateSnapshot(BaseModel):
    """服务端权威运行状态；连接建立和每次命令后完整返回。"""

    model_config = ConfigDict(extra="forbid")

    pipeline: str
    subtitle: str
    mic_muted: bool
    persona: str | None
    voice: str
    duplex_mode: DuplexMode
    session_started_at: str | None
    mode: RuntimeMode = RuntimeMode.ASSISTANT
    pcm_owner: PCMOwner = PCMOwner.NONE
    active_meeting_id: str | None = None
    meeting_state: MeetingStatus | None = None
    meeting_started_at: str | None = None
    storage: StorageHealth = StorageHealth.OK
    runtime_revision: int = Field(default=0, ge=0)


class ErrorCode(StrEnum):
    INVALID_COMMAND = "invalid_command"
    INVALID_PAYLOAD = "invalid_payload"
    COMMAND_FAILED = "command_failed"
    SERVICE_UNAVAILABLE = "service_unavailable"
    STORAGE_UNAVAILABLE = "storage_unavailable"
    TRANSCRIPTION_UNAVAILABLE = "transcription_unavailable"
    MODE_CONFLICT = "mode_conflict"
    MEETING_NOT_ACTIVE = "meeting_not_active"
    FINALIZATION_TIMEOUT = "finalization_timeout"


class CommandResponse(BaseModel):
    """命令确认；失败仅暴露稳定错误码与安全消息。"""

    model_config = ConfigDict(extra="forbid")

    request_id: str
    cmd: str
    ok: bool
    state: RuntimeStateSnapshot
    error_code: ErrorCode | None = None
    message: str | None = None
    contract_version: Literal["1"] | None = None
    error: dict[str, Any] | None = None


class RuntimeStateEvent(BaseModel):
    """控制连接建立时主动发送的状态握手。"""

    model_config = ConfigDict(extra="forbid")

    event: Literal["state", "runtime_state"] = "runtime_state"
    state: RuntimeStateSnapshot
    contract_version: Literal["1"] | None = None


def parse_command(payload: Mapping[str, Any]) -> ControlCommand:
    """严格解析控制命令，拒绝未知命令、缺失字段和额外字段。"""
    return _COMMAND_ADAPTER.validate_python(dict(payload))
